_drafts/assert.md

```diff
@@ -0,0 +1,605 @@
+---
+title: "Using Asserts in Embedded Systems"
+description:
+  "Asserts are a quick and proven way to prevent bugs in embedded systems. An
+  optimal assert implementation both maximizes the context around the assert
+  while minimizing code size and RAM usage."
+author: tyler
+image: /img/debugging-asserts/assert.png
+---
+
+The use of asserts is one of the best ways to find bugs, unintended behavior,
+programmatic errors, and to learn when systems are no longer 100% functional and
+need to be reset. If instrumented correctly, an assert can give a developer
+context about when and where in the code an issue took place. Despite the
+numerous benefits, the practice of using asserts in firmware is not common or
+agreed upon.
+
+<!-- excerpt start -->
+
+By using asserts proactively in embedded systems on debug and production builds,
+developers can both prevent bugs before shipping and quickly surface and fix
+them after shipping. Proper assert handling is also the safest way to handle
+issues and undefined behavior that occur in production. In this post, we'll go
+over best practices with asserts, when to use asserts, and then come up with a
+production ready custom assert implementation for an ARM Cortex-M device, all
+while keeping the code size usage to a minimum.
+
+<!-- excerpt end -->
+
+_Like Interrupt? [Subscribe](http://eepurl.com/gpRedv) to get our latest posts
+straight to your mailbox_
+
+{:.no_toc}
+
+## Table of Contents
+
+<!-- prettier-ignore -->
+* auto-gen TOC:
+{:toc}
+
+## Asserts in Embedded Systems
+
+The use of asserts is controversial in the embedded space, **especially** in
+released or production builds. On top of adhering to the ANSI C standard of
+`NDEBUG` disabling `assert()`s, many of the popular embedded platforms allow one
+to
+[compile](https://github.com/aws/amazon-freertos/blob/master/vendors/nordic/nRF5_SDK_15.2.0/components/libraries/util/nrf_assert.h#L78-L115)
+[out](https://github.com/espressif/esp-idf/blob/master/components/esp_common/include/esp_err.h#L104-L124)
+[error](https://github.com/ARMmbed/mbed-os/blob/master/platform/mbed_error.h#L191-L202)
+[handling](https://github.com/zephyrproject-rtos/zephyr/blob/master/include/sys/__assert.h#L34-L89).
+
+In my opinion, disabling asserts for any reason other than to check how much
+codesize asserts take up in the system is more than likely a **reckless**
+endeavor, especially if they are used correctly.
+
+### Article Disclaimers
+
+- This post is about runtime asserting, not compile time asserting, such as
```

Comment:
Compile time asserts would make for a short and good subsection, and they don't justify an article on their own. Just a thought.